site/bulletin_board/data/posts.js

```diff
@@ -51,6 +52,66 @@ posts.retrieve = (id, userId, callback) => {
   });
 };
 
+
+posts.retrieveperuser = (userId, callback) => {
+  var sql = `
+    SELECT
+      Posts.id AS id,
+      Posts.title,
+      Author.username AS author,
+      Author.id AS authorid,
+      Posts.date,
+      Posts.body AS body,
+      PostUpvotes.post_id IS NOT NULL AS liked,
+      '/posts/' + Posts.id AS url
+    FROM
+      Posts
+      INNER JOIN Users AS Author ON Posts.user_id = Author.id
+      LEFT OUTER JOIN PostUpvotes ON PostUpvotes.post_id = Posts.id AND PostUpvotes.user_id = ?
+    WHERE
+    Author.id = ?
+    ORDER BY
+      date DESC
+  `;
+  db.all(sql, [ userId,userId ], (err, rows) => {
+    if (err) {
+      callback([]);
+      return;
+    }
+    callback(rows);
+  });
+}
+
+
+
+posts.post_liked = (userId, callback) => {
+  var sql = `
+    SELECT
+      Posts.id AS id,
+      Posts.title,
+      Author.username AS author,
+      Author.id AS authorid,
+      Posts.date,
+      Posts.body AS body,
+      PostUpvotes.post_id IS NOT NULL AS liked,
+      '/posts/' + Posts.id AS url
+    FROM
+      Posts
+      INNER JOIN Users AS Author ON Posts.user_id = Author.id
+      LEFT OUTER JOIN PostUpvotes ON PostUpvotes.post_id = Posts.id AND PostUpvotes.user_id = ?
+    WHERE
+    liked = 1
+    ORDER BY
+      date DESC
```

Comment:
`LEFT OUTER JOIN` is good to use when you want the table you're joining to be "optional". In this case, we only ever want Posts that have a PostUpvote from `userId`, so using an `INNER JOIN` will let the SQL engine be smarter about how it processes your query (and makes your intention much clearer too... I found this query confusing to understand at first).